c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -423,128 +437,267 @@ class CachingAllocatorConfig {
     return *s_instance;
   }
 
-  void parseArgs(const char* env) {
-    // If empty, set the default values
-    m_max_split_size = std::numeric_limits<size_t>::max();
-    m_roundup_power2_divisions = 0;
-    m_roundup_bypass_threshold = std::numeric_limits<size_t>::max();
-    m_garbage_collection_threshold = 0;
+  void parseArgs(const char* env);
 
-    if (env == nullptr) {
-      return;
+ private:
+  CachingAllocatorConfig()
+      : m_max_split_size(std::numeric_limits<size_t>::max()),
+        m_garbage_collection_threshold(0) {
+    m_roundup_power2_divisions.assign(Native::kRoundUpPowerOfTwoIntervals, 0);
+  }
+
+  void lexArgs(const char* env, std::vector<std::string>& config);
+  void consumeToken(
+      const std::vector<std::string>& config,
+      size_t& i,
+      const char c);
+  void parseMaxSplitSize(const std::vector<std::string>& config, size_t& i);
+  void parseGarbageCollectionThreshold(
+      const std::vector<std::string>& config,
+      size_t& i);
+  void parseRoundUpPower2Divisions(
+      const std::vector<std::string>& config,
+      size_t& i);
+  void parseAllocatorConfig(
+      const std::vector<std::string>& config,
+      size_t& i,
+      bool& used_cudaMallocAsync);
+
+  std::atomic<size_t> m_max_split_size;
+  std::vector<size_t> m_roundup_power2_divisions;
+  std::atomic<double> m_garbage_collection_threshold;
+};
+
+void CachingAllocatorConfig::lexArgs(
+    const char* env,
+    std::vector<std::string>& config) {
+  std::string buf;
+  for (size_t i = 0; i < strlen(env); i++) {
+    if (env[i] == ',' || env[i] == ':' || env[i] == '[' || env[i] == ']') {
+      if (!buf.empty()) {
+        config.emplace_back(buf);
+        buf = "";
+      }
+      config.emplace_back(std::string(1, env[i]));
+    } else if (
+        isalpha(env[i]) || isdigit(env[i]) || env[i] == '>' || env[i] == '_' ||
```

Comment:
To make this easier for new options int he future, and since the tokens have to be checked for validity later anyway:
```
else if (env[i] != ' ') {
  buf.emplace_back(env[i])
}
```